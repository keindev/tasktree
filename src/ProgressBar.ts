import figures from 'figures';

import { TaskStatus } from './Task.js';
import { IndicationType, TextSeparator, Theme } from './Theme.js';

/** Progress in percent by default */
export enum Progress {
  Default = -1,
  Start = 0,
  End = 100,
}

/** ProgressBar output template */
export enum TemplateToken {
  /** The progress bar itself */
  Bar = ':bar',
  /** Current tick number */
  Current = ':current',
  /** Total ticks */
  Total = ':total',
  /**  */
  // Time elapsed in seconds
  Elapsed = ':elapsed',
  /** Completion percentage */
  Percent = ':percent',
  /** Estimated completion time in seconds */
  ETA = ':eta',
  /** Rate of ticks per second */
  Rate = ':rate',
}

/** ProgressBar display options */
export interface IProgressBarOptions {
  /** Option to add badge */
  badges?: boolean;
  /** Option to clear the bar on completion */
  clear?: boolean;
  /** Completion character */
  completeChar?: string;
  /** Current completed index */
  current?: number;
  /** Option to add gradient to pending bar */
  gradient?: boolean;
  /** Incomplete character */
  incompleteChar?: string;
  /** Total number of ticks to complete */
  total?: number;
  /** The displayed width of the progress bar defaulting to total */
  width?: number;
}

/** User template tokens values */
export interface IProgressBarToken {
  [key: string]: string;
}

/**
 * To add a Progress Bar to your task, use {@link Task.bar}  method accepting the string pattern and progress bar parameters.
 *
 * ```javascript
 * const progress = new task.bar([template, options]);
 * ```
 */
export class ProgressBar implements Required<Omit<IProgressBarOptions, 'current'>> {
  static readonly MAX_PERCENT = 100;
  static readonly MAX_POINT_POSITION = 1;
  static readonly MAX_RATIO = 1;
  static readonly MIN_PERCENT = 0;
  static readonly MIN_POINT_POSITION = 0;
  static readonly MIN_RATIO = 0;
  static readonly TICK = 1;
  static readonly TIME_DIMENSION = 1000;

  /** Option to add badge */
  readonly badges: boolean = true;
  /** Option to clear the bar on completion */
  readonly clear: boolean = false;
  /** Completion character */
  readonly completeChar = figures.square;
  /** Option to add gradient to pending bar */
  readonly gradient: boolean = true;
  /** Incomplete character */
  readonly incompleteChar = figures.square;
  /**
   * Output template
   *
   * @default `:bar :rate/bps :percent :eta/s`
   */
  readonly template: string;
  /** Total number of ticks to complete */
  readonly total = Progress.End;
  /** The displayed width of the progress bar defaulting to total */
  readonly width: number = 20;

  #current = Progress.Start;
  #end: number | undefined;
  readonly #start = new Date().getTime();
  #status = TaskStatus.Pending;
  #tokens: Map<TemplateToken | string, string> = new Map();

  constructor(template?: string, options?: IProgressBarOptions) {
    this.template =
      template ||
      Theme.join(
        TextSeparator.Empty,
        TemplateToken.Bar,
        `${TemplateToken.Rate}/bps`,
        TemplateToken.Percent,
        `${TemplateToken.ETA}s`
      );

    if (typeof options === 'object') {
      if (options.current) this.#current = options.current;
      if (options.total) this.total = options.total;
      if (options.width) this.width = options.width;
      if (options.completeChar) this.completeChar = options.completeChar;
      if (options.incompleteChar) this.incompleteChar = options.incompleteChar;

      this.clear = !!options.clear;
    }
  }

  /** Ratio between `current` value and `total` value */
  get ratio(): number {
    return Math.min(Math.max(this.#current / this.total, ProgressBar.MIN_RATIO), ProgressBar.MAX_RATIO);
  }

  /** Current percent of completion */
  get percent(): number {
    return Math.floor(this.ratio * ProgressBar.MAX_PERCENT);
  }

  /** Elapsed time from the beginning of progress, in milliseconds */
  get elapsed(): number {
    return (this.#end || new Date().getTime()) - this.#start;
  }

  /** Rate of progress */
  get rate(): number {
    return this.#current / (this.elapsed / ProgressBar.TIME_DIMENSION);
  }

  /** Progress ETA (estimated time of arrival) */
  get ETA(): number {
    return this.percent === ProgressBar.MAX_PERCENT
      ? ProgressBar.MIN_PERCENT
      : this.elapsed * (this.total / this.#current - ProgressBar.MAX_RATIO);
  }

  /** Start `Date` in milliseconds */
  get start(): number {
    return this.#start;
  }

  /** End `Date` in milliseconds if progress is an ended */
  get end(): number | undefined {
    return this.#end;
  }

  /** `true` if progress is complete */
  get isCompleted(): boolean {
    return this.#current >= this.total || !!this.#end;
  }

  /** Completes progress and marks it as successful */
  complete(): void {
    this.#current = this.total;
    this.#status = TaskStatus.Completed;
    this.#end = new Date().getTime();
  }

  /** Stops the progress and marks it as failed */
  fail(): void {
    if (!this.isCompleted) {
      this.#status = TaskStatus.Failed;
      this.#end = new Date().getTime();
    }
  }

  /** Render output string with Theme */
  render(theme: Theme): string {
    const length = Math.round(this.width * this.ratio);
    const type = Theme.type(this.#status);

    let result = this.template
      .replace(TemplateToken.Current, this.#current.toString())
      .replace(TemplateToken.Total, this.total.toString())
      .replace(TemplateToken.Percent, `${this.percent.toFixed(ProgressBar.MIN_POINT_POSITION)}%`)
      .replace(TemplateToken.ETA, (this.ETA / ProgressBar.TIME_DIMENSION).toFixed(ProgressBar.MAX_POINT_POSITION))
      .replace(TemplateToken.Rate, Math.round(this.rate).toString())
      .replace(
        TemplateToken.Elapsed,
        (this.elapsed / ProgressBar.TIME_DIMENSION).toFixed(ProgressBar.MAX_POINT_POSITION)
      )
      .replace(
        TemplateToken.Bar,
        Theme.join(
          TextSeparator.Empty,
          this.getBlocks(theme, type, length),
          theme.paint(TextSeparator.Empty.padStart(this.width - length, this.incompleteChar), IndicationType.Subtask)
        )
      );

    this.#tokens.forEach((value: string, key: string): void => {
      result = result.replace(`:${key}`, value);
    });

    return this.badges ? Theme.join(TextSeparator.Space, result, theme.badge(type)) : result;
  }

  /** Stops the progress and marks it as skipped */
  skip(): void {
    if (!this.isCompleted) {
      this.#status = TaskStatus.Skipped;
      this.#end = new Date().getTime();
    }
  }

  /**
   * Increases current progress on step value
   * @param step - Value by which the current progress will increase
   * @param tokens - Add custom tokens by adding a `{'name': value}` object parameter to your method
   * @example
   * ```javascript
   * const bar = new Progress(':bar template with custom :token');
   *
   * bat.tick(10, { token: 100 });
   * ```
   */
  tick(step?: number, tokens?: IProgressBarToken): ProgressBar {
    this.#current = Math.min(this.total, this.#current + (step || ProgressBar.TICK));
    this.#tokens = typeof tokens === 'object' ? new Map(Object.entries(tokens)) : new Map();

    if (this.isCompleted) this.complete();

    return this;
  }

  private getBlocks(theme: Theme, type: IndicationType, length: number): string {
    const str = TextSeparator.Empty.padStart(length, this.completeChar);

    return !this.isCompleted && this.gradient
      ? theme.gradient(str, { position: this.ratio, begin: Theme.type(this.#status), end: IndicationType.Success })
      : theme.paint(str, type);
  }
}
