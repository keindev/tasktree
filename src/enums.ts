export enum Status {
    Pending = 0,
    Completed = 1,
    Failed = 2,
    Skipped = 3,
}

export enum Level {
    Default = 0,
    Step = 1,
    Stride = 3,
}

export enum Progress {
    Default = -1,
    Start = 0,
    End = 100,
}

export enum Token {
    // the progress bar itself
    Bar = ':bar',
    // current tick number
    Current = ':current',
    // total ticks
    Total = ':total',
    // time elapsed in seconds
    Elapsed = ':elapsed',
    // completion percentage
    Percent = ':percent',
    // estimated completion time in seconds
    ETA = ':eta',
    // rate of ticks per second
    Rate = ':rate',
}

export enum Indent {
    Default = 2,
    Empty = 0,
}

export enum Type {
    Default = 'default',
    Active = 'active',
    Success = 'success',
    Skip = 'skip',
    Error = 'error',
    Message = 'message',
    Info = 'info',
    Warning = 'warning',
    Subtask = 'subtask',
    List = 'list',
    Dim = 'dim',
}

export enum Color {
    Default = '',
    Active = '#4285f4',
    Success = '#00c851',
    Skip = '#ff8800',
    Error = '#ff4444',
    Message = '#2e2e2e',
    Info = '#33b5e5',
    Warning = '#ffbb33',
    Subtask = '#2e2e2e',
    List = '#4285f4',
    Dim = '#838584',
}

export enum Badge {
    Default = '',
    Error = '[fail]',
    Skip = '[skip]',
}
