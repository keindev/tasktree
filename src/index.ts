export { default } from './TaskTree.js';
