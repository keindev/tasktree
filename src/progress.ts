import * as Figures from 'figures';
import * as Enums from './enums';
import { Theme } from './theme';
import { Options, Token } from './types';

export class Progress {
    public static TICK = 1;
    public static TIME_DIMENSION = 1000;
    public static MAX_POINT_POSITION = 1;
    public static MIN_POINT_POSITION = 0;
    public static MIN_PERCENT = 0;
    public static MAX_PERCENT = 100;
    public static MIN_RATIO = 0;
    public static MAX_RATIO = 1;

    public readonly total: number = Enums.Progress.End;
    public readonly width: number = 20;
    public readonly completeBlock: string = Figures.square;
    public readonly incompleteBlock: string = Figures.square;
    public readonly clear: boolean = false;
    public readonly template: string;

    private current: number = Enums.Progress.Start;
    private start: number = new Date().getTime();
    private tokens: Map<string, string> = new Map();

    public constructor(template?: string, options?: Options) {
        this.template =
            template ||
            Theme.join(Enums.Token.Bar, `${Enums.Token.Rate}/bps`, Enums.Token.Percent, `${Enums.Token.ETA}s`);

        if (typeof options === 'object') {
            if (options.current) this.current = options.current;
            if (options.total) this.total = options.total;
            if (options.width) this.width = options.width;
            if (options.complete) this.completeBlock = options.complete;
            if (options.incomplete) this.incompleteBlock = options.incomplete;

            this.clear = !!options.clear;
        }
    }

    public getRatio(): number {
        return Math.min(Math.max(this.current / this.total, Progress.MIN_RATIO), Progress.MAX_RATIO);
    }

    public getPercent(): number {
        return Math.floor(this.getRatio() * Progress.MAX_PERCENT);
    }

    public getElapsed(): number {
        return new Date().getTime() - this.start;
    }

    public getRate(): number {
        return this.current / (this.getElapsed() / Progress.TIME_DIMENSION);
    }

    public getETA(): number {
        return this.getPercent() === Progress.MAX_PERCENT
            ? Progress.MIN_PERCENT
            : this.getElapsed() * (this.total / this.current - Progress.MAX_RATIO);
    }

    public isCompleted(): boolean {
        return this.current >= this.total;
    }

    public tick(step?: number, tokens?: Token): boolean {
        const { total } = this;

        this.current = Math.min(total, this.current + (step || Progress.TICK));
        this.tokens = typeof tokens === 'object' ? new Map(Object.entries(tokens)) : new Map();

        return this.isCompleted();
    }

    public complete(): void {
        this.current = this.total;
    }

    public render(theme: Theme): string {
        const length = Math.round(this.width * this.getRatio());

        let result = this.template
            .replace(Enums.Token.Current, this.current.toString())
            .replace(Enums.Token.Total, this.total.toString())
            .replace(Enums.Token.Percent, `${this.getPercent().toFixed(Progress.MIN_POINT_POSITION)}%`)
            .replace(Enums.Token.ETA, (this.getETA() / Progress.TIME_DIMENSION).toFixed(Progress.MAX_POINT_POSITION))
            .replace(Enums.Token.Rate, Math.round(this.getRate()).toString())
            .replace(
                Enums.Token.Elapsed,
                (this.getElapsed() / Progress.TIME_DIMENSION).toFixed(Progress.MAX_POINT_POSITION)
            )
            .replace(
                Enums.Token.Bar,
                [
                    theme.paint(Theme.EMPTY.padStart(length, this.completeBlock), Enums.Type.Success),
                    theme.paint(Theme.EMPTY.padStart(this.width - length, this.incompleteBlock), Enums.Type.Subtask),
                ].join(Theme.EMPTY)
            );

        this.tokens.forEach(
            (value: string, key: string): void => {
                result = result.replace(`:${key}`, value);
            }
        );

        return result;
    }
}
