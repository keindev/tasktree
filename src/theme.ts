import chalk from 'chalk';
import convert from 'color-convert';
import elegantSpinner from 'elegant-spinner';
import * as Figures from 'figures';
import * as Types from './types';
import * as Enums from './enums';
import { Task } from './task';
import { Progress } from './progress';

const frame = elegantSpinner();

export class Theme {
    public static SEPARATOR = '\n';
    public static EMPTY = '';
    public static SPACE = ' ';
    public static INDENT = '  ';

    private colors: Map<Enums.Type, Types.Color> = new Map();
    private figures: Map<Enums.Type, Types.Figure> = new Map();
    private badges: Map<Enums.Type, Types.Badge> = new Map();

    public constructor(options?: Types.Theme) {
        if (options) {
            const { colors, figures, badges } = this;
            const list = new Map(Object.entries(options));
            const set = (key: Enums.Type, color?: Types.Color, figure?: Types.Figure, badge?: Types.Badge): void => {
                if (color) colors.set(key, color);
                if (figure) figures.set(key, figure);
                if (badge) badges.set(key, badge);
            };

            Object.values(Enums.Type).forEach(
                (key: Enums.Type): void => {
                    const option = list.get(key);

                    if (Array.isArray(option)) {
                        set(key, ...option);
                    } else if (typeof option === 'object') {
                        const { color, figure, badge } = option;

                        set(key, color, figure, badge);
                    }
                }
            );
        }
    }

    public static type(status: Enums.Status, isList: boolean = false): Enums.Type {
        let type: Enums.Type;

        switch (status) {
            case Enums.Status.Completed:
                type = Enums.Type.Success;
                break;
            case Enums.Status.Skipped:
                type = Enums.Type.Skip;
                break;
            case Enums.Status.Failed:
                type = Enums.Type.Error;
                break;
            case Enums.Status.Pending:
            default:
                type = isList ? Enums.Type.List : Enums.Type.Active;
                break;
        }

        return type;
    }

    public static join(separator: string, ...text: string[]): string {
        return text.filter((value): boolean => !!value.length).join(separator);
    }

    public static dye(str: string, color: Types.Color): string {
        return color ? chalk.hex(color)(str) : str;
    }

    public static indent(count: number, ...text: string[]): string {
        return `${Theme.INDENT.padStart(count * Enums.Indent.Default)}${Theme.join(Theme.SPACE, ...text)}`;
    }

    private static getValueBy<T>(map: Map<Enums.Type, T>, type: Enums.Type, getDefault: () => T): T {
        let result = map.get(type);

        if (!result) result = getDefault();

        return result;
    }

    public getColor(type: Enums.Type): Types.Color {
        const { colors } = this;

        return Theme.getValueBy(
            this.colors,
            type,
            (): string => {
                if (type === Enums.Type.Active) return Enums.Color.Active;
                if (type === Enums.Type.Success) return Enums.Color.Success;
                if (type === Enums.Type.Skip) return Enums.Color.Skip;
                if (type === Enums.Type.Error) return Enums.Color.Error;
                if (type === Enums.Type.Message) return Enums.Color.Message;
                if (type === Enums.Type.Info) return Enums.Color.Info;
                if (type === Enums.Type.Warning) return Enums.Color.Warning;
                if (type === Enums.Type.Exception) return Enums.Color.Exception;
                if (type === Enums.Type.Subtask) return Enums.Color.Subtask;
                if (type === Enums.Type.List) return Enums.Color.List;
                if (type === Enums.Type.Dim) return Enums.Color.Dim;

                return colors.get(Enums.Type.Default) || Enums.Color.Default;
            }
        );
    }

    public paint(str: string, type: Enums.Type): string {
        return Theme.dye(str, this.getColor(type));
    }

    public gradient(str: string, gradient: Types.Gradient): string {
        const begin = convert.hex.rgb(this.getColor(gradient.begin));
        const end = convert.hex.rgb(this.getColor(gradient.end));
        const w = gradient.position * 2 - 1;
        const w1 = (w + 1) / 2.0;
        const w2 = 1 - w1;

        return Theme.dye(
            str,
            convert.rgb.hex([
                Math.round(end[0] * w1 + begin[0] * w2),
                Math.round(end[1] * w1 + begin[1] * w2),
                Math.round(end[2] * w1 + begin[2] * w2),
            ])
        );
    }

    public figure(type: Enums.Type): Types.Figure {
        const { figures } = this;
        const figure = Theme.getValueBy(
            figures,
            type,
            (): string => {
                if (type === Enums.Type.Active) return frame();
                if (type === Enums.Type.Success) return Figures.tick;
                if (type === Enums.Type.Skip) return Figures.arrowDown;
                if (type === Enums.Type.Error) return Figures.cross;
                if (type === Enums.Type.Message) return Figures.line;
                if (type === Enums.Type.Info) return Figures.info;
                if (type === Enums.Type.Warning) return Figures.warning;
                if (type === Enums.Type.Exception) return Figures.arrowRight;
                if (type === Enums.Type.Subtask) return Figures.pointerSmall;
                if (type === Enums.Type.List) return Figures.pointer;

                return figures.get(Enums.Type.Default) || this.figure(Enums.Type.Subtask);
            }
        );

        return figure ? this.paint(figure, type) : figure;
    }

    public badge(type: Enums.Type): Types.Badge {
        const { badges } = this;
        const badge = Theme.getValueBy(
            badges,
            type,
            (): string => {
                if (type === Enums.Type.Error) return Enums.Badge.Error;
                if (type === Enums.Type.Skip) return Enums.Badge.Skip;

                return badges.get(Enums.Type.Default) || Enums.Badge.Default;
            }
        );

        return badge ? this.paint(badge, Enums.Type.Dim) : badge;
    }

    public title(task: Task, level: number): string {
        const type = Theme.type(task.getStatus(), task.isList());
        const badge = this.badge(type);
        const figure = this.figure(type);
        let prefix = Theme.EMPTY;

        if (level) prefix = task.isList() ? this.figure(Enums.Type.Subtask) : this.figure(Enums.Type.Default);

        return Theme.indent(level, prefix, figure, task.getText(), badge);
    }

    public errors(errors: string[], level: number): string[] {
        const type = Enums.Type.Error;
        const sublevel = level + Enums.Level.Stride;

        return errors.map(
            (text): string => {
                const [error, ...lines] = text.split(Theme.SEPARATOR);
                const title = Theme.join(Theme.SPACE, this.figure(Enums.Type.Message), this.figure(type), error.trim());

                return Theme.join(
                    Theme.SEPARATOR,
                    Theme.indent(level + Enums.Level.Step, this.paint(title, type)),
                    ...lines.map((line): string => Theme.indent(sublevel, this.paint(line.trim(), Enums.Type.Dim)))
                );
            }
        );
    }

    public messages(list: string[], type: Enums.Type, level: number): string[] {
        const figure = this.figure(type);
        const sign = this.figure(Enums.Type.Message);
        const indent = level + Enums.Level.Step;

        return list.map((text): string => Theme.indent(indent, sign, figure, text));
    }

    public bars(list: Progress[], level: number): string[] {
        const figure = this.figure(Enums.Type.Subtask);

        return list
            .filter((bar): boolean => !bar.isCompleted() || !bar.clear)
            .map((bar): string => Theme.indent(level, figure, bar.render(this)));
    }
}
