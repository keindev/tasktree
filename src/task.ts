import chalk from 'chalk';
import { TaskTree } from './tasktree';
import { Theme } from './theme';
import * as Enums from './enums';
import { Progress } from './progress';
import { Options } from './types';

let uid = 0;

export class Task {
    private uid: number;
    private text: string;
    private status: Enums.Status;
    private bars: Progress[] = [];
    private subtasks: Task[] = [];
    private logs: Set<string> = new Set();
    private errors: string[] = [];
    private warnings: Set<string> = new Set();

    public constructor(text: string, status: Enums.Status = Enums.Status.Pending) {
        this.uid = ++uid;
        this.text = text;
        this.status = status;
    }

    public id(): number {
        return this.uid;
    }

    public getText(): string {
        return this.text;
    }

    public getStatus(): Enums.Status {
        return this.status;
    }

    public getActive(): Task {
        const { subtasks } = this;
        const subtask = subtasks[subtasks.length - 1];
        let task: Task = this;

        if (subtask && subtask.isPending()) task = subtask.getActive();

        return task;
    }

    public isPending(): boolean {
        return this.status === Enums.Status.Pending;
    }

    public havePendingSubtasks(): boolean {
        return !!this.subtasks.filter((task): boolean => task.isPending()).length;
    }

    public haveSubtasks(): boolean {
        return !!this.subtasks.length;
    }

    public add(text: string, status: Enums.Status = Enums.Status.Pending): Task {
        const isCompleted = !this.isPending();
        const task = new Task(text, isCompleted ? Enums.Status.Failed : status);

        this.subtasks.push(task);

        if (isCompleted) this.fail(`Task is already complete, can't add new subtask [${task.id()}]`);

        return task;
    }

    public bar(template?: string, options?: Options): Progress {
        const isCompleted = !this.isPending();
        const bar = new Progress(template, isCompleted ? { total: Enums.Progress.End } : options);

        this.bars.push(bar);

        if (isCompleted) this.fail("Task is already complete, can't add new progress bar");

        return bar;
    }

    public clear(): void {
        this.subtasks = [];
        this.bars = [];
    }

    public complete(text?: string, clear: boolean = false): Task {
        if (this.havePendingSubtasks()) this.fail('Subtasks is not complete.');

        this.update(Enums.Status.Completed, text, clear);
        this.bars = this.bars.filter((bar): boolean => {
            bar.complete();

            return !bar.clear;
        });

        return this;
    }

    public skip(text?: string, clear: boolean = false): Task {
        this.update(Enums.Status.Skipped, text, clear);

        return this;
    }

    public fail(text?: string, clear: boolean = false): never {
        this.update(Enums.Status.Failed, text, clear);

        TaskTree.tree().exit(Enums.ExitCode.Error);

        throw new Error(text);
    }

    public error(error?: string | Error, fail?: boolean): Task {
        const { errors } = this;

        if (typeof error === 'string') errors.push(error);
        if (error instanceof Error && error.stack) errors.push(error.stack);
        if (fail) this.fail();

        return this;
    }

    public log(text: string): Task {
        if (this.isPending()) this.logs.add(text);

        return this;
    }

    public warn(text: string): Task {
        if (this.isPending()) this.warnings.add(text);

        return this;
    }

    public render(theme: Theme, level = Enums.Level.Default): string[] {
        const type = level ? Enums.Type.Dim : Enums.Type.Default;
        const rows = [
            theme.title(this, level),
            ...theme.bars(this.bars, level + Enums.Level.Step),
            ...theme.errors(this.errors, level),
            ...theme.messages([...this.warnings], Enums.Type.Warning, level),
            ...theme.messages([...this.logs], Enums.Type.Info, level),
        ];

        this.subtasks.forEach((task): void => {
            rows.push(...task.render(theme, level + Enums.Level.Step));
        });

        return rows.map((row): string => theme.paint(row, type));
    }

    private update(status: Enums.Status, text?: string, clear: boolean = false): void {
        if (this.isPending()) {
            if (text) this.text = text;
            if (clear) this.clear();

            this.status = status;
        } else {
            this.error(`Task is already complete (${chalk.bold(this.status.toString())})`);
            this.status = Enums.Status.Failed;
        }
    }
}
