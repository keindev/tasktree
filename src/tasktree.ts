import { UpdateManager } from 'stdout-update';
import { Task } from './task';
import * as Types from './types';
import { Theme } from './theme';
import { ExitCode } from './enums';

export class TaskTree {
    public static TIMEOUT = 100;
    private static instance: TaskTree;

    private handle: NodeJS.Timeout | undefined;
    private tasks: Task[];
    private theme: Theme;
    private manager: UpdateManager;
    private silence: boolean = false;

    private constructor(theme?: Types.Theme) {
        this.tasks = [];
        this.theme = new Theme(theme);
        this.manager = UpdateManager.getInstance();
    }

    public static tree(theme?: Types.Theme): TaskTree {
        if (!TaskTree.instance) {
            TaskTree.instance = new TaskTree(theme);
        }

        return TaskTree.instance;
    }

    public start(silence?: boolean): TaskTree {
        this.silence = !!silence;
        this.tasks = [];

        if (!this.handle && !this.silence) {
            this.manager.hook();
            this.handle = setInterval((): void => {
                this.log();
            }, TaskTree.TIMEOUT);
        }

        return this;
    }

    public stop(): TaskTree {
        if (this.handle) {
            clearInterval(this.handle);

            this.log();
            this.manager.unhook();
            this.handle = undefined;
        }

        return this;
    }

    public exit(code: ExitCode = ExitCode.Success): void {
        if (!this.silence) {
            this.stop();
            process.exit(code);
        }
    }

    public add(text: string): Task {
        const { tasks } = this;
        let task = tasks[tasks.length - 1];

        if (task && task.isPending()) {
            task = task.getActive();
            task = task.add(text);
        } else {
            tasks.push((task = new Task(text)));
        }

        return task;
    }

    public render(): string[] {
        return this.tasks.reduce<string[]>((acc, task): string[] => acc.concat(task.render(this.theme)), []);
        // return this.tasks.map((task): string[] => task.render(this.theme)).flat();
    }

    private log(): void {
        this.manager.update(this.render());
    }
}
