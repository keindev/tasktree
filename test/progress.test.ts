import stripAnsi from 'strip-ansi';
import { Progress } from '../src/progress';
import { Theme } from '../src/theme';

describe('Progress', (): void => {
    const $template = ':bar :percent :etas :custom';
    const $theme = new Theme();

    it('Default', (): void => {
        const step = 1;
        const bar = new Progress($template, { total: step * 2 });

        expect(bar.getPercent()).toBe(Progress.MIN_PERCENT);
        expect(bar.getRatio()).toBe(Progress.MIN_RATIO);
        expect(bar.getStart()).toBeTruthy();
        expect(bar.getEnd()).toBeFalsy();
        expect(bar.getElapsed()).toBeTruthy();
        expect(bar.getRate()).toBeFalsy();
        expect(bar.getETA()).toBe(Infinity);
        expect(bar.isCompleted()).toBeFalsy();

        bar.tick();

        expect(bar.getPercent()).toBe(Progress.MAX_PERCENT / 2);
        expect(bar.getRatio()).toBe(Progress.MAX_RATIO / 2);
        expect(bar.getStart()).toBeTruthy();
        expect(bar.getEnd()).toBeFalsy();
        expect(bar.getElapsed()).toBeTruthy();
        expect(bar.getRate()).toBeTruthy();
        expect(bar.getETA()).toBeTruthy();
        expect(bar.isCompleted()).toBeFalsy();

        bar.tick(step, {
            custom: 'test',
        });

        expect(bar.getPercent()).toBe(Progress.MAX_PERCENT);
        expect(bar.getRatio()).toBe(Progress.MAX_RATIO);
        expect(bar.getStart()).toBeTruthy();
        expect(bar.getEnd()).toBeTruthy();
        expect(bar.getElapsed()).toBeTruthy();
        expect(bar.getRate()).toBeTruthy();
        expect(bar.getETA()).toBeFalsy();
        expect(bar.isCompleted()).toBeTruthy();
        expect(stripAnsi(bar.render($theme))).toMatchSnapshot();
    });

    describe('Statuses', (): void => {
        let $bar: Progress;
        let $before: number;

        beforeEach((): void => {
            $bar = new Progress($template);
            $before = new Date().getTime();

            $bar.tick($bar.total / 2);
        });

        afterEach((): void => {
            expect($bar.isCompleted()).toBeTruthy();
            expect($bar.getEnd()).toBeGreaterThanOrEqual($before);
            expect($bar.getEnd()).toBeLessThanOrEqual(new Date().getTime());
        });

        it('Complete', (): void => {
            $bar.complete();

            expect($bar.getPercent()).toBe(Progress.MAX_PERCENT);
            expect(stripAnsi($bar.render($theme))).toMatchSnapshot();
        });

        it('Skip', (): void => {
            $bar.skip();

            expect($bar.getPercent()).toBe(Progress.MAX_PERCENT / 2);
            expect(stripAnsi($bar.render($theme))).toMatchSnapshot();
        });

        it('Fail', (): void => {
            $bar.fail();

            expect($bar.getPercent()).toBe(Progress.MAX_PERCENT / 2);
            expect(stripAnsi($bar.render($theme))).toMatchSnapshot();
        });
    });
});
