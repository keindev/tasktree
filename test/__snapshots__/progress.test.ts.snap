// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Default 1`] = `"▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 100% 0.0s test"`;

exports[`Progress Statuses Complete 1`] = `"▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 100% 0.0s :custom"`;

exports[`Progress Statuses Fail 1`] = `"▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 50% 0.0s :custom [fail]"`;

exports[`Progress Statuses Skip 1`] = `"▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 50% 0.0s :custom [skip]"`;
